### PR TITLE
only do a multipart s3 copy when above the regular copy limit

### PR DESCRIPTION
**Pull request by icewind1991**
Multipart copy seems problematic with some backends so avoid it when possible.

### CODE REVIEW

**Review by solracsf — COMMENTED** [review]
Why don't test if MultiPart is supported? Something like:

```
if ($this->getObjectStore() instanceof IObjectStoreMultiPartUpload) {
   // multipart
} else {
  // not multipart
}
```

File: apps/files_external/tests/Storage/Amazons3MultiPartTest.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		parent::tearDown();
+	}
+
+	public function testStat() {
+		$this->markTestSkipped('S3 doesn\'t update the parents folder mtime');
+	}
+
+	public function testHashInFileName() {
+		$this->markTestSkipped('Localstack has a bug with hashes in filename');
+	}
```
Comment (come-nc) [inline apps/files_external/tests/Storage/Amazons3MultiPartTest.php:67]: This adds skipped tests? Why?
Output of phpunit is full of skipped tests and it’s making it hard to see failures. It’s best to use groups and exclude non-relevant tests.

**Review by come-nc — APPROVED** [review]

File: apps/files_external/tests/Storage/Amazons3MultiPartTest.php (hunk 1 above)
Comment (icewind1991) [reply apps/files_external/tests/Storage/Amazons3MultiPartTest.php:67]: These tests overwrite the tests from the parent class

**Review by juliusknorr — APPROVED** [review]
Seems good, do you know what s3 providers/solutions enforce such a limit?

**Comment by juliushaertl** [comment]
> Why don't test if MultiPart is supported?

It might need a different interface, I'm not sure if all providers that support multipart **upload** also support multipart **copy**

**Comment by Ziyann** [comment]
For reference, I'm also affected by this issue, using Google Cloud Storage (but haven't tested this patch yet). Some built-in solution would be greatly appreciated, either a configurable limit like `copySizeLimit` here (which we could set to a very high value, thus never triggering multipart copy), or even better, some enable/disable switches for the various operations, because right now I'm manually reverting https://github.com/nextcloud/server/pull/40577 on production after upgrades.
There are many S3 providers with different features supported, so I think it would be great to have as many configuration options as possible (avoiding manual files changes on production).

**Comment by juliushaertl** [comment]
Pushed an additional commit to be able to fully turn off multipart copy for s3 providers not supporting it at all.

**Pull request merged**